File: java/com/joshmanisdabomb/lcc/item/ComputingBlockItem.java
package com.joshmanisdabomb.lcc.item;

import com.joshmanisdabomb.lcc.computing.ComputingModule;
import com.joshmanisdabomb.lcc.registry.LCCBlocks;
import net.minecraft.block.Block;
import net.minecraft.item.*;
import net.minecraft.util.NonNullList;

import java.util.Map;

public class ComputingBlockItem extends BlockItem implements TintedItem {

    private final DyeColor color;
    private final ComputingModule.Type module;

    public ComputingBlockItem(ComputingModule.Type module, DyeColor color, Properties properties) {
        super(LCCBlocks.computing, properties);
        this.color = color;
        this.module = module;
    }

    @Override
    public void fillItemGroup(ItemGroup group, NonNullList<ItemStack> items) {
        if (this.isInGroup(group)) {
            items.add(new ItemStack(this));
        }
    }

    @Override
    public String getTranslationKey() {
        return this.getDefaultTranslationKey();
    }

    @Override
    public void addToBlockToItemMap(Map<Block, Item> map, Item i) {

    }

    @Override
    public void removeFromBlockToItemMap(Map<Block, Item> map, Item i) {

    }

    @Override
    public int getItemTintColor(ItemStack stack, int tintIndex) {
        return this.color.getColorValue();
    }

    public DyeColor getColor() {
        return this.color;
    }

    public ComputingModule.Type getModule() {
        return this.module;
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/item/block/ComputingBlockItem.kt
package com.joshmanisdabomb.lcc.item.block

import com.joshmanisdabomb.lcc.abstracts.computing.module.ComputerModule
import com.joshmanisdabomb.lcc.block.entity.ComputingBlockEntity
import com.joshmanisdabomb.lcc.block.entity.DungeonTableBlockEntity
import com.joshmanisdabomb.lcc.directory.LCCBlocks
import com.joshmanisdabomb.lcc.lib.item.DefaultedColoredItem
import com.joshmanisdabomb.lcc.trait.LCCItemTrait
import net.minecraft.block.Block
import net.minecraft.entity.ItemEntity
import net.minecraft.item.BlockItem
import net.minecraft.item.Item
import net.minecraft.item.ItemGroup
import net.minecraft.item.ItemStack
import net.minecraft.util.collection.DefaultedList
import net.minecraft.util.math.Direction

class ComputingBlockItem(val module: ComputerModule, settings: Settings) : PlasticBlockItem(LCCBlocks.computing, settings), DefaultedColoredItem, LCCItemTrait {

    override fun getTranslationKey() = getOrCreateTranslationKey()

    override fun appendBlocks(map: MutableMap<Block, Item>, item: Item) = Unit

    override fun appendStacks(group: ItemGroup, stacks: DefaultedList<ItemStack>) {
        if (!this.isIn(group)) return
        stacks.add(ItemStack(this))
    }

    fun getComputingHalf(be: ComputingBlockEntity, stack: ItemStack, direction: Direction, top: Boolean): ComputingBlockEntity.ComputingHalf {
        val half = be.ComputingHalf(this.module, direction, this.getTintColor(stack), top)
        if (stack.hasCustomName()) half.customName = stack.name
        return half
    }

}